lib/libc/sys/ktrace.2
.\" Copyright (c) 1993 Regents of the University of California.
.\" All rights reserved.
.\"
.\" %sccs.include.redist.man%
.\"
.\"     @(#)ktrace.2	6.1 (Berkeley) 6/1/93
.\"
.Dd 
.Dt KTRACE 2
.Os BSD 4
.Sh NAME
.Nm ktrace
.Nd process tracing
.Sh SYNOPSIS
.Fd #include <sys/types.h>
.Fd #include <sys/ktrace.h>
.Ft int
.Fn ktrace "const char *tracefile" "int ops" "int trpoints" "int pid"
.Sh DESCRIPTION
The
.Fn ktrace
function enables or disables tracing of one or more processes.
Users may only trace their own processes.
Only the super-user can trace setuid or setgid programs.
.Pp
The
.Ar tracefile
gives the pathname of the file to be used for tracing.
The file must exist and be writable by the calling process.
All trace records are always appended to the file,
so the file must be truncated to zero length to discard
previous trace data.
If tracing points are being disabled (see KTROP_CLEAR below),
.Ar tracefile
may be NULL.
.Pp
The
.Nm ops
parameter specifies the requested ktrace operation.
The defined operations are:
.Bl -column KTRFLAG_DESCENDXXX -offset indent
.It KTROP_SET	Enable trace points specified in Ar trpoints .
.It KTROP_CLEAR	Disable trace points specified in Ar trpoints .
.It KTROP_CLEARFILE	Stop all tracing.
.It KTRFLAG_DESCEND	The tracing change should apply to the
specified process and all its current children.
.El
.Pp
The
.Nm trpoints
parameter specifies the trace points of interest.
The defined trace points are:
.Bl -column KTRFAC_SYSCALLXXX -offset indent
.It KTRFAC_SYSCALL	Trace system calls.
.It KTRFAC_SYSRET	Trace return values from system calls.
.It KTRFAC_NAMEI	Trace name lookup operations.
.It KTRFAC_GENIO	Trace all I/O (note that this option can
generate much output).
.It KTRFAC_PSIG	Trace posted signals.
.It KTRFAC_CSW	Trace context switch points.
.It KTRFAC_INHERIT	Inherit tracing to future children.
.El
.Pp
Each tracing event outputs a record composed of a generic header
followed by a trace point specific structure.
The generic header is:
.Bd -literal
struct ktr_header {
	int	ktr_len;		/* length of buf */
	short	ktr_type;		/* trace record type */
	pid_t	ktr_pid;		/* process id */
	char	ktr_comm[MAXCOMLEN+1];	/* command name */
	struct	timeval ktr_time;	/* timestamp */
	caddr_t	ktr_buf;
};
.Ed
.Pp
The
.Nm ktr_len
field specifies the length of the 
.Nm ktr_type
data that follows this header.
The
.Nm ktr_pid
and
.Nm ktr_comm
fields specify the process and command generating the record.
The
.Nm ktr_time
field gives the time (with microsecond resolution)
that the record was generated.
The
.Nm ktr_buf
is an internal kernel pointer and is not useful.
.Pp
The generic header is followed by 
.Nm ktr_len
bytes of a
.Nm ktr_type
record.
The type specific records are defined in the
.Pa <sys/ktrace.h>
include file.
.Sh RETURN VALUES
On successful completion a value of 0 is returned.
Otherwise, a value of -1 is returned and
.Va errno
is set to show the error.
.Sh ERRORS
.Fn Ktrace
will fail if:
.Bl -tag -width ENAMETOOLONGAA
.It Bq Er ENOTDIR
A component of the path prefix is not a directory.
.It Bq Er EINVAL
The pathname contains a character with the high-order bit set.
.It Bq Er ENAMETOOLONG
A component of a pathname exceeded 255 characters,
or an entire path name exceeded 1023 characters.
.It Bq Er ENOENT
The named tracefile does not exist.
.It Bq Er EACCES
Search permission is denied for a component of the path prefix.
.It Bq Er ELOOP
Too many symbolic links were encountered in translating the pathname.
.It Bq Er EIO
An I/O error occurred while reading from or writing to the file system.
.El
.Sh SEE ALSO
.Xr ktrace 1 ,
.Xr kdump 1
.Sh HISTORY
A
.Nm ktrace
function call first appeared in
.Bx 4.4 .


usr.bin/ktrace/ktrace.c
/*
 * Copyright (c) 1988 The Regents of the University of California.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms are permitted
 * provided that the above copyright notice and this paragraph are
 * duplicated in all such forms and that any documentation,
 * advertising materials, and other materials related to such
 * distribution and use acknowledge that the software was developed
 * by the University of California, Berkeley.  The name of the
 * University may not be used to endorse or promote products derived
 * from this software without specific prior written permission.
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED
 * WARRANTIES OF MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.
 */

#ifndef lint
char copyright[] =
"@(#) Copyright (c) 1988 The Regents of the University of California.\n\
 All rights reserved.\n";
#endif /* not lint */

#ifndef lint
static char sccsid[] = "@(#)ktrace.c	1.4 (Berkeley) 3/7/90";
#endif /* not lint */

#include "ktrace.h"

#define USAGE \
 "usage: ktrace [-acid] [-f trfile] [-t trops] [-p pid] [-g pgid]\n\
	trops: c = syscalls, n = namei, g = generic-i/o, a = everything\n\
	ktrace -C (clear everthing)\n"
	

char	*tracefile = DEF_TRACEFILE;
int	append, clear, descend, inherit;

main(argc, argv)
	char *argv[];
{
	extern int optind;
	extern char *optarg;
	int facs = ALL_FACS;
	int ops = 0;
	int pid = 0;
	int ch;

	while ((ch = getopt(argc,argv,"Cacdp:g:if:t:")) != EOF)
		switch((char)ch) {
		case 'C':
			clear = 2;
			break;
		case 'c':
			clear = 1;
			break;
		case 'd':
			ops |= KTRFLAG_DESCEND;
			break;
		case 't':
			facs = getfacs(optarg);
			if (facs < 0) {
				fprintf(stderr, 
				    "ktrace: unknown facility in %s\n",
			 	     optarg);
				exit(1);
			}
			break;
		case 'p':
			pid = atoi(optarg);
			break;
		case 'g':
			pid = -atoi(optarg);
			break;
		case 'i':
			inherit++;
			break;
		case 'f':
			tracefile = optarg;
			break;
		case 'a':
			append++;
			break;
		default:
			fprintf(stderr,"usage: \n",*argv);
			exit(-1);
		}
	argv += optind, argc -= optind;
	
	if (inherit)
		facs |= KTRFAC_INHERIT;
	if (clear) {			/* untrace something */
		if (clear == 2) {	/* -C */
			ops = KTROP_CLEAR | KTRFLAG_DESCEND;
			pid = 1;
		} else {
			ops |= pid ? KTROP_CLEAR : KTROP_CLEARFILE;
		}
		if (ktrace(tracefile, ops, facs, pid) < 0) {
			perror("ktrace");
			exit(1);
		}
		exit(0);
	}

	if (pid == 0 && !*argv) {	/* nothing to trace */
		fprintf(stderr, USAGE);
		exit(1);
	}
			
	close(open(tracefile, O_WRONLY | O_CREAT, 0666));
	if (!append)
		close(open(tracefile, O_WRONLY | O_TRUNC));
	if (!*argv) {
		if (ktrace(tracefile, ops, facs, pid) < 0) {
			perror("ktrace");
			exit(1);
		}
	} else {
		pid = getpid();
		if (ktrace(tracefile, ops, facs, pid) < 0) {
			perror("ktrace");
			exit(1);
		}
		execvp(argv[0], &argv[0]);
		perror("ktrace: exec failed");
	}
	exit(0);
}
